### PR TITLE
UI: 3D Weapon Preview

### PR DESCRIPTION
**Pull request by TimGoll**
Adds a 3D weapon preview that could be used in the shop rework.

Things done:

- [ ] Trigger attack animation [[discord]](https://discord.com/channels/565105920414318602/567617926991970306/1199019906050572329) (probably too much, so it is scrapped for now)
- [x] Support weapons such as boomerang and melon mine [[discord]](https://discord.com/channels/565105920414318602/567617926991970306/1199087704470716426)
- [x] reduce player opacity on hover to highlight weapon

image

Demovideo:
https://github.com/TTT-2/TTT2/assets/13639408/98e4bf81-6ca7-420d-ac57-fbaeee3c4bbc

### CODE REVIEW

File: gamemodes/terrortown/gamemode/shared/sh_include.lua [hunk 1]
```diff
@@ -117,7 +117,8 @@ TTTFiles = {
 	cl_vskin__vgui__dsearchbar = {file = "cl_vskin/vgui/dsearchbar_ttt2.lua", on = "client"},
 	cl_vskin__vgui__dsubmenulist = {file = "cl_vskin/vgui/dsubmenulist_ttt2.lua", on = "client"},
 	cl_vskin__vgui__dinfoitem = {file = "cl_vskin/vgui/dinfoitem_ttt2.lua", on = "client"},
-	cl_vskin__vgui__dprofilepanel = {file = "cl_vskin/vgui/dprofilepanel_ttt2.lua", on = "client"}
+	cl_vskin__vgui__dprofilepanel = {file = "cl_vskin/vgui/dprofilepanel_ttt2.lua", on = "client"},
+	cl_vskin__vgui__dweaponpreview = {file = "cl_vskin/vgui/dweaponpreview_ttt2.lua", on = "client"}
```
Comment (ZenBre4ker) [inline gamemodes/terrortown/gamemode/shared/sh_include.lua:121]:
```suggestion
	cl_vskin__vgui__dweaponpreview = {file = "cl_vskin/vgui/dweaponpreview_ttt2.lua", on = "client"},
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -26,6 +26,9 @@ All notable changes to TTT2 will be documented here. Inspired by [keep a changel
 - Icon for gameplay menu
 - Icon for accessibility menu
 - Icon for `Voice & Volume` menu
+- Added a new vgui element: `DWeaponPreview_TTT2` to render a player with their equipped weapon (by @TimGoll)
+  - Supports any normal weapon that has a `.HoldType` and a `.WorldModel`
+  - Supports any weapon that is made with the SWEP Construction Kit (bommerang, melonmine, ...)
```
Comment (ZenBre4ker) [inline CHANGELOG.md:31]:
```suggestion
  - Supports any weapon that is made with the SWEP Construction Kit (boomerang, melonmine, ...)
```

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+
+			local matrix = Matrix()
+			matrix:Scale(worldModelData.size)
+			wep:EnableMatrix("RenderMultiply", matrix)
+
+			if worldModelData.material == "" then
+				wep:SetMaterial("")
+			elseif wep:GetMaterial() ~= worldModelData.material then
+				wep:SetMaterial( worldModelData.material )
+			end
```
Comment (ZenBre4ker) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:272]: This seems like it does just this:
```suggestion
			wep:SetMaterial( worldModelData.material )
```

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+
+			if worldModelData.skin and worldModelData.skin ~= wep:GetSkin() then
+				wep:SetSkin(worldModelData.skin)
+			end
+
+			if worldModelData.bodygroup then
+				for k, v in pairs(worldModelData.bodygroup) do
+					if wep:GetBodygroup(k) ~= v then
+						wep:SetBodygroup(k, v)
+					end
```
Comment (ZenBre4ker) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:282]: not a fan of just k,v. even key, value is not saying much.

**Review by ZenBre4ker — APPROVED** [review]
Mostly LGTM.

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua (hunk 3 above)
Comment (TimGoll) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:272]: Nope, that's not the same

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+
+	local ang = self.aLookAngle or (self.vLookatPos - self.vCamPos):Angle()
+
+	cam.Start3D(self.vCamPos, ang, self.fFOV, xBaseStart, yBaseStart, w, h, 5, self.farZ)
+		render.SuppressEngineLighting(true)
+		render.SetLightingOrigin(ply:GetPos())
+		render.SetColorModulation(self.colColor.r / 255, self.colColor.g / 255, self.colColor.b / 255)
+
+		for i = 0, 6 do
+			local col = self.directionalLight[i]
```
[the remark this answers is not in the document]
Comment (TimGoll) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:306]: As this code is copied from original TTT, I hope this works as intended: https://github.com/Facepunch/garrysmod/blob/87aa68cb8da3d70e6e7265bc61ca222c55e3f8e5/garrysmod/lua/vgui/dmodelpanel.lua#L139-L144

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua (hunk 3 above)
Comment (ZenBre4ker) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:272]: But the first statement sets it to `""` if material is `""`. 
And the second one sets the material always if it is not the material.
Therefore they are both always setting worldModelData.material.

Comment (TimGoll) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:272]: I agree with you, I just thought it has a reason this is done that way. If issues arise, we can point it to that change

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua (hunk 4 above)
Comment (TimGoll) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua]: Renamed it to `bodygroup`/`value` according to [this](https://wiki.facepunch.com/gmod/Entity:SetBodygroup)

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua (hunk 5 above)
[the remark this answers is not in the document]
Comment (ZenBre4ker) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:302]: Yeah apparently it is just enums
https://wiki.facepunch.com/gmod/Enums/BOX

After you read this, we can resolve it.

[the remark this answers is not in the document]
Comment (TimGoll) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:302]: I'll add this as a comment so that it is clear in the future

[the remark this answers is not in the document]
Comment (TimGoll) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:302]: It probably is a C programmer's bug then, because they assumed that the upper limit is not included. I'll keep it that way though

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+
+			if worldModelData.bodygroup then
+				for bodygroup, value in pairs(worldModelData.bodygroup) do
+					if wep:GetBodygroup(bodygroup) ~= value then
+						wep:SetBodygroup(bodygroup, value)
+					end
+				end
+			end
+		end
+	end -- end of SWEP Construction Kit support
```
Comment (saibotk) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:282]: replace with the renderer from PR #1325

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+
+	if not IsValid(ply) then return end
+
+	local w, h = self:GetSize()
+	local xBaseStart, yBaseStart = self:LocalToScreen(0, 0)
+
+	local xLimitStart, yLimitStart = xBaseStart, yBaseStart
+	local xLimitEnd, yLimitEnd = self:LocalToScreen(self:GetWide(), self:GetTall())
+
+	local curparent = self
```
Comment (saibotk) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:224]: `currentParent`

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+function PANEL:Init()
+	self.lastPaint = 0
+	self.directionalLight = {}
+	self.farZ = 4096
+
+	self:SetContentAlignment(5)
+
+	self:SetTall(22)
+	self:SetMouseInputEnabled(true)
+	self:SetKeyboardInputEnabled(true)
```
Comment (saibotk) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:71]: What do we need keyboard input for?

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+function PANEL:SetDirectionalLight(iDirection, color)
+	self.directionalLight[iDirection] = color
+end
+
+---
+-- @param string model
+-- @realm client
+function PANEL:SetPlayerModel(model)
+	-- set the entity
+	local cEnt = ClientsideModel(model, RENDERGROUP_OTHER)
```
Comment (saibotk) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:129]: no abbreviations

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+	if IsValid(self.data.ply) then
+		self.data.ply:Remove()
+	end
+	if IsValid(self.data.wep) then
+		self.data.wep:Remove()
+	end
+end
+
+---
+-- @param number iDirections
```
Comment (saibotk) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:117]: where does this number come from?

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+	-- and special handling has to be used to render the weapon
+	if wep.WElements then
+		self.data.specialModelHandeling = true
+
+		for name, data in pairs(wep.WElements) do
+			if data.type ~= "Model" then continue end
+
+			self.data.worldModelData[name] = data
+
+			local cEnt = ClientsideModel(data.model, RENDERGROUP_OTHER)
```
Comment (saibotk) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:176]: no abbreviations

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+
+			if not IsValid(cEnt) then continue end
+
+			self.data.worldModels[name] = cEnt
+		end
+	else
+		local cEnt = ClientsideModel(wep.WorldModel, RENDERGROUP_OTHER)
```
Comment (saibotk) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:183]: no abbreviations

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+
+		if not IsValid(cEnt) then return end
+
+		cEnt:SetNoDraw(true)
+		cEnt:SetIK(false)
+
+		cEnt:SetParent(self.data.ply)
+		cEnt:AddEffects(EF_BONEMERGE)
```
Comment (saibotk) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:191]: Why is this needed? Seems like this could use a comment

**Review by saibotk — CHANGES_REQUESTED** [review]

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua (hunk 8 above)
Comment (TimGoll) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:71]: nothing, this was a copy paste error

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua (hunk 10 above)
Comment (TimGoll) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:115]: image

This is taken from the other 3d model UI renderers in Gmod

**Comment by TimGoll** [comment]
done @saibotk

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua (hunk 6 above)
Comment (saibotk) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:283]: Mhh yeah lets halt it and just use the other function once that is merged, which is making good progress

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua (hunk 10 above)
Comment (saibotk) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:115]: Yes but the documentation for this parameter could mention the ENUM this is coming from instead of `iDirections`

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dprofilepanel_ttt2.lua [hunk 14]
```diff
@@ -177,7 +177,8 @@ function PANEL:DrawModel(x, y, w, h)
     render.SetColorModulation(self.colColor.r / 255, self.colColor.g / 255, self.colColor.b / 255)
     render.SetBlend((self:GetAlpha() / 255) * (self.colColor.a / 255))
 
-    for i = 0, 6 do
+        -- iterates over the model lighting enum: https://wiki.facepunch.com/gmod/Enums/BOX
+        for i = 0, 6 do
```
Comment (Histalek) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dprofilepanel_ttt2.lua:181]: this indentation change doesn't seem correct

Comment (TimGoll) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dprofilepanel_ttt2.lua:181]: still working on this PR!

**Comment by TimGoll** [comment]
I think this PR is now ready as well

File: lua/ttt2/libraries/weaponrenderer.lua [hunk 15]
```diff
@@ -353,3 +361,50 @@ function weaponrenderer.Render(wep, elements, boneEntity)
...
+-- positions so that the view model tracks the hands.
+-- @param Entity wep The weapon whose view model should be rendered
+-- @param table elements The elements of the view model
+-- @param Entity viewModel The player's view model
+-- @realm client
+function weaponrenderer.RenderViewModel(wep, elements, viewModel)
+    if not elements then
+        return
+    end
+
```
Comment (saibotk) [inline lua/ttt2/libraries/weaponrenderer.lua:376]: I think such a check does not belong in here:

1. It silently does nothing, which is wanted here but weird for a module function
2. all other parameters are not checked

This is something that should be handled where this is called.

File: lua/ttt2/libraries/weaponrenderer.lua [hunk 16]
```diff
@@ -353,3 +361,50 @@ function weaponrenderer.Render(wep, elements, boneEntity)
...
+-- @param[opt] Player owner The owner entity of the weapon, binds the model to their hands
+-- @realm client
+function weaponrenderer.RenderWoldModel(wep, wepModel, elements, owner)
+    if wep.ShowDefaultWorldModel then
+        wepModel:DrawModel()
+    end
+
+    if not elements then
+        return
+    end
```
Comment (saibotk) [inline lua/ttt2/libraries/weaponrenderer.lua:398]: same here

File: lua/ttt2/libraries/weaponrenderer.lua [hunk 17]
```diff
@@ -353,3 +361,50 @@ function weaponrenderer.Render(wep, elements, boneEntity)
...
+---
+-- Renders the wold model if valid wold model elements are provided. It also renders the default
+-- world model of the weapon if enabled
+-- @param Entity wep The weapon whose world model should be rendered
+-- @param Entity wepModel The weapon mode whose world model should be rendered, in most cases
+-- identical to the first parameter
+-- @param table elements The elements of the world model
+-- @param[opt] Player owner The owner entity of the weapon, binds the model to their hands
+-- @realm client
+function weaponrenderer.RenderWoldModel(wep, wepModel, elements, owner)
```
Comment (saibotk) [inline lua/ttt2/libraries/weaponrenderer.lua:391]:
```suggestion
function weaponrenderer.RenderWorldModel(wep, wepModel, elements, owner)
```

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua [hunk 18]
```diff
@@ -0,0 +1,350 @@
...
+            self.colAmbientLight.b / 255
+        )
+        render.SetBlend((self:GetAlpha() / 255) * (self.colColor.a / 255))
+    end
+
+    ply:DrawModel()
+
+    render.SetBlend((self:GetAlpha() / 255) * (self.colColor.a / 255))
+
+    weaponrenderer.RenderWoldModel(
```
Comment (saibotk) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:294]:
```suggestion
    weaponrenderer.RenderWorldModel(
```

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua [hunk 19]
```diff
@@ -0,0 +1,350 @@
...
+    self:SetText("")
+
+    self.data = {
+        ply = nil,
+        wep = nil,
+        wepClass = "",
+        HoldType = "normal",
+        worldModelData = {},
+        worldModels = {},
+        specialModelHandeling = false,
```
Comment (saibotk) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:101]: These are both unused now and can be removed along with code that uses them.

**Review by saibotk — CHANGES_REQUESTED** [review]
I like these changes a lot! Good job!

File: lua/ttt2/libraries/weaponrenderer.lua (hunk 16 above)
Comment (TimGoll) [reply lua/ttt2/libraries/weaponrenderer.lua:398]: Not sure how to do it here, hm.
This function also draws the normal weapon world model and the custom one can only be drawn if it is set up. Therefore an external check doesn't work.

I can change it for the view model, but that makes the handling between these two functions incosistent

**Comment by TimGoll** [comment]
Done @saibotk

> I like these changes a lot! Good job!

Took me long enough to figure out how to handle these ClientsideModels

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+
+    -- before storing the ent, make sure that a possible old ent
+    -- is removed because clientside models are not garbage collected
+    if IsValid(self.data.wep) then
+        self.data.wep:Remove()
+    end
+
+    -- make sure it is reset in case different weapon was used before
+    self.data.wepClass = cls
+
```
Comment (saibotk) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dweaponpreview_ttt2.lua:181]:
```suggestion
```

**Review by saibotk — CHANGES_REQUESTED** [review]

**Review by saibotk — APPROVED** [review]
